File: ipc-codegen/echo_example/ts/src/echo_server.ts
/**
 * Echo IPC server (TypeScript) — uses the GENERATED handleRequest (framing,
 * dispatch, and error wrapping) over the @aztec/ipc-runtime UDS transport.
 * Usage: npx tsx echo_server.ts --socket /tmp/echo.sock
 */
import { UdsIpcServer } from "@aztec/ipc-runtime";
import { handleRequest } from "./generated/server.js";
import type { Handler } from "./generated/server.js";
import type {
  EchoBytes,
  EchoBytesResponse,
  EchoAliases,
  EchoAliasesResponse,
  EchoBlobs,
  EchoBlobsResponse,
  EchoFail,
  EchoFailResponse,
  EchoFields,
  EchoFieldsResponse,
  EchoNested,
  EchoNestedResponse,
} from "./generated/api_types.js";

const args = process.argv.slice(2);
const socketIdx = args.indexOf("--socket");
const socketPath = socketIdx >= 0 ? args[socketIdx + 1] : undefined;
if (!socketPath) {
  console.error("Usage: echo_server.ts --socket <path>");
  process.exit(1);
}

const handler: Handler = {
  async echoBytes(cmd: EchoBytes): Promise<EchoBytesResponse> {
    return { data: cmd.data };
  },
  async echoFields(cmd: EchoFields): Promise<EchoFieldsResponse> {
    return { a: cmd.a, b: cmd.b, name: cmd.name };
  },
  async echoNested(cmd: EchoNested): Promise<EchoNestedResponse> {
    return { inner: cmd.inner };
  },
  async echoAliases(cmd: EchoAliases): Promise<EchoAliasesResponse> {
    return {
      treeId: cmd.treeId,
      hash: cmd.hash,
      maybeHash: cmd.maybeHash,
      hashes: cmd.hashes,
    };
  },
  async echoBlobs(cmd: EchoBlobs): Promise<EchoBlobsResponse> {
    return { maybeData: cmd.maybeData, parts: cmd.parts };
  },
  async echoFail(cmd: EchoFail): Promise<EchoFailResponse> {
    throw new Error(cmd.message);
  },
};

async function main() {
  await UdsIpcServer.listen(socketPath!, (_clientId, requestBytes) =>
    handleRequest(handler, requestBytes),
  );
  console.error(`ipc-server(ts): listening on ${socketPath}`);
}

main().catch((e) => {
  console.error(`echo_server(ts): FAILED: ${e.message ?? e}`);
  process.exit(1);
});

File: ipc-codegen/echo_example/zig/src/echo_server.zig
//! Echo IPC server (Zig) — uses the ipc-runtime Zig binding for transport
//! and the GENERATED Dispatcher for framing, dispatch, and error wrapping.
//! Usage: echo_server --socket /tmp/echo.sock
const std = @import("std");
const ipc_runtime = @import("ipc_runtime");
const types = @import("generated/echo_types.zig");
const echo_server = @import("generated/echo_server.zig");

const EchoHandler = struct {
    // Handlers are asynchronous: they produce their result via respond.ok(...) /
    // respond.err(...) (synchronously here; an async transport could defer).
    pub fn bytes(self: *EchoHandler, cmd: types.EchoBytes, respond: *echo_server.Responder(types.EchoBytesResponse)) void {
        _ = self;
        respond.ok(.{ .data = cmd.data });
    }

    pub fn fields(self: *EchoHandler, cmd: types.EchoFields, respond: *echo_server.Responder(types.EchoFieldsResponse)) void {
        _ = self;
        respond.ok(.{ .a = cmd.a, .b = cmd.b, .name = cmd.name });
    }

    pub fn nested(self: *EchoHandler, cmd: types.EchoNested, respond: *echo_server.Responder(types.EchoNestedResponse)) void {
        _ = self;
        respond.ok(.{ .inner = cmd.inner });
    }

    pub fn aliases(self: *EchoHandler, cmd: types.EchoAliases, respond: *echo_server.Responder(types.EchoAliasesResponse)) void {
        _ = self;
        respond.ok(.{
            .tree_id = cmd.tree_id,
            .hash = cmd.hash,
            .maybe_hash = cmd.maybe_hash,
            .hashes = cmd.hashes,
        });
    }

    pub fn blobs(self: *EchoHandler, cmd: types.EchoBlobs, respond: *echo_server.Responder(types.EchoBlobsResponse)) void {
        _ = self;
        respond.ok(.{ .maybe_data = cmd.maybe_data, .parts = cmd.parts });
    }

    pub fn fail(self: *EchoHandler, cmd: types.EchoFail, respond: *echo_server.Responder(types.EchoFailResponse)) void {
        _ = self;
        respond.err(cmd.message);
    }
};

const EchoDispatcher = echo_server.Dispatcher(EchoHandler);

pub fn main() !void {
    var args = std.process.args();
    _ = args.next();
    var socket_path: ?[:0]const u8 = null;
    while (args.next()) |arg| {
        if (std.mem.eql(u8, arg, "--socket")) {
            socket_path = args.next();
        }
    }
    const path = socket_path orelse {
        std.debug.print("Usage: echo_server --socket <path>\n", .{});
        return error.InvalidArgument;
    };

    var handler = EchoHandler{};
    var dispatcher = EchoDispatcher.init(&handler);

    var server = try ipc_runtime.Server.fromPath(path);
    defer server.deinit();
    server.installDefaultSignalHandlers();
    try server.listen();
    std.debug.print("ipc-server(zig): listening on {s}\n", .{path});

    server.run(*EchoDispatcher, &dispatcher, EchoDispatcher.handleRequest);
}
